test_conformance/api/test_api_consistency.cpp

```diff
@@ -0,0 +1,479 @@
+//
+// Copyright (c) 2020 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "testBase.h"
+#include "harness/testHarness.h"
+
+const char* test_kernel = R"CLC(
+__kernel void test(__global int* dst) {
+    dst[0] = 0;
+}
+)CLC";
+
+int test_consistency_svm(cl_device_id deviceID, cl_context context,
+                         cl_command_queue queue, int num_elements)
+{
+    // clGetDeviceInfo, passing CL_DEVICE_SVM_CAPABILITIES:
+    // May return 0, indicating that device does not support Shared Virtual
+    // Memory.
+    int error;
+
+    const size_t allocSize = 16;
+    clMemWrapper mem;
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+
+    cl_device_svm_capabilities svmCaps = 0;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_SVM_CAPABILITIES,
+                            sizeof(svmCaps), &svmCaps, NULL);
+    test_error(error, "Unable to query CL_DEVICE_SVM_CAPABILITIES");
+
+    if (svmCaps == 0)
+    {
+        mem =
+            clCreateBuffer(context, CL_MEM_READ_WRITE, allocSize, NULL, &error);
+        test_error(error, "Unable to create test buffer");
+
+        error = create_single_kernel_helper(context, &program, &kernel, 1,
+                                            &test_kernel, "test");
+        test_error(error, "Unable to create test kernel");
+
+        // clGetMemObjectInfo, passing CL_MEM_USES_SVM_POINTER
+        // Returns CL_FALSE if no devices in the context associated with
+        // memobj support Shared Virtual Memory.
+        cl_bool usesSVMPointer;
+        error =
+            clGetMemObjectInfo(mem, CL_MEM_USES_SVM_POINTER,
+                               sizeof(usesSVMPointer), &usesSVMPointer, NULL);
+        test_error(error, "Unable to query CL_MEM_USES_SVM_POINTER");
+        if (usesSVMPointer != CL_FALSE)
+        {
+            log_error("CL_DEVICE_SVM_CAPABILITIES returned 0 but "
+                      "CL_MEM_USES_SVM_POINTER did not return CL_FALSE\n");
+            return TEST_FAIL;
+        }
+
+        // Check that the SVM APIs can be called.
+
+        // Returns NULL if no devices in context support Shared Virtual Memory.
+        void* ptr0 = clSVMAlloc(context, CL_MEM_READ_WRITE, allocSize, 0);
+        void* ptr1 = clSVMAlloc(context, CL_MEM_READ_WRITE, allocSize, 0);
```

Comment:
Why two calls? Because two is better than one?

Reply:
I wanted two pointers to for two reasons: first, to have one SVM pointer to "fill" and a different SVM pointer to "copy" to, and second, to have one SVM pointer to free via `clSVMFree` and another SVM pointer to free via `clEnqueueSVMFree`.

Since we're enforcing errors from these functions I suppose this isn't strictly necessary and we can pass bogus arguments to each of these APIs, but is still feels like good hygiene to have a reasonably valid code sequence.

Reply:
Ok, I understand the reasoning now. Fine as is.